### PR TITLE
send retry-after until the apiserver is ready

### LINKED ISSUE #104342: GC might delete objects during kube-apiserver startup
#### What happened:
GC may delete objects during startup while the apiserver has not fully initialized yet.

One potential case where it can happen: CRDs are available through informers 
- kube-apiserver is starting, its informers have not synced yet
- GC sends a query (GET a CR)
- since the informers have not synced yet, GC may get a `404`
- This may potentially lead GC to delete objects 

#### What you expected to happen:
`kube-apiserver` should respond with a `Retry-After` until it has fully initialized. GC will get a `429` with a `Retry-After` response header in this case.

#### How to reproduce it (as minimally and precisely as possible):

With an HA cluster it's hard to reproduce.

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Introduce a new server run option `startup-send-retry-after-until-ready` 
```
If true, incoming request(s) will be rejected with a '429' status code and a 'Retry-After' response 
header until the apiserver has initialized. This option ensures that the system stays consistent 
even when requests arrive at the server before it has been initialized. 
```

The default value is set to `false` to maintain current behavior.

#### Which issue(s) this PR fixes:

Fixes #104342

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go [hunk 1]
```diff
@@ -69,14 +69,26 @@ func WithRetryAfter(handler http.Handler, when ShouldRespondWithRetryAfterFunc)
 }
 
 // NewShouldRespondWithRetryAfterFunc returns a ShouldRespondWithRetryAfterFunc
-func NewShouldRespondWithRetryAfterFunc(shutdownSendRetryAfter bool, shutdownCh <-chan struct{}) ShouldRespondWithRetryAfterFunc {
-	if !shutdownSendRetryAfter {
+func NewShouldRespondWithRetryAfterFunc(shutdownSendRetryAfter, startupSendRetryAfterUntilReady bool, shutdownCh, readyCh <-chan struct{}) ShouldRespondWithRetryAfterFunc {
+	switch {
+	case shutdownSendRetryAfter && startupSendRetryAfterUntilReady:
```
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:74]: maybe we can take a different approach if this looks hard to follow.

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 2]
```diff
@@ -800,7 +800,8 @@ func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
 	handler = genericapifilters.WithCacheControl(handler)
 	handler = genericfilters.WithHSTS(handler, c.HSTSDirectives)
 	if c.ShutdownSendRetryAfter {
-		handler = genericfilters.WithRetryAfter(handler, c.lifecycleSignals.AfterShutdownDelayDuration.Signaled())
+		shouldRetryAfterFn := genericfilters.NewShouldRespondWithRetryAfterFunc(c.ShutdownSendRetryAfter, c.lifecycleSignals.AfterShutdownDelayDuration.Signaled())
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config.go]: the NewShouldRespondWithRetryAfterFunc helper should be a prive func here turning the signal into a predicate `func() bool`.

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go [hunk 3]
```diff
@@ -62,25 +62,40 @@ type shouldRespondWithRetryAfterFunc func() (*retryAfterParams, bool)
...
+// NewShouldRespondWithRetryAfterFunc returns a ShouldRespondWithRetryAfterFunc
+func NewShouldRespondWithRetryAfterFunc(shutdownSendRetryAfter bool, shutdownCh <-chan struct{}) ShouldRespondWithRetryAfterFunc {
+	if !shutdownSendRetryAfter {
+		return func() (*retryAfterParams, bool) {
+			return nil, false
+		}
+	}
+
+	return newShutdownRetryAfterFunc(shutdownCh)
+}
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:92]: see comment in config.go. This func does not add value. The signal->bool conversion belongs into a local helper in config.go.

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 4]
```diff
@@ -799,8 +869,8 @@ func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
 	handler = genericapifilters.WithWarningRecorder(handler)
 	handler = genericapifilters.WithCacheControl(handler)
 	handler = genericfilters.WithHSTS(handler, c.HSTSDirectives)
-	if c.ShutdownSendRetryAfter {
-		handler = genericfilters.WithRetryAfter(handler, c.lifecycleSignals.AfterShutdownDelayDuration.Signaled())
+	if shouldRespondWithRetryAfterFn, ok := c.shouldAddWithRetryAfterFilter(); ok {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:872]: this conditional seems unnecessary complicated. Why not apply the filter always and make the predicate clever enough?

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/config.go]: comment still applies. Make it a local helper. This is a trivial func part of the plumbing here.

**Comment by sttts** [comment]
/lgtm
/approve

Would have squashed the commits. The intermediate state after the first does not have much value.
